index.js

```diff
@@ -62,40 +82,93 @@ app.get("/username/:name", (req, res) => {
     res.statusCode = 200;
     res.send("name can be used");
   }
-  console.log(names);
 });
 
 app.post("/username", (req, res) => {
-  if (isNameUsed(req.body)) {
+  if (isNameUsed(req.body.name)) {
     res.statusCode = 409;
     res.send("name is already in use");
   } else {
-    saveName(req.body); 
+    saveName(req.body.name, req.body.key);
     res.statusCode = 200;
     res.send("user created");
   }
 });
 
+app.post("/redeem", (req, res) => {
+  if (giftcodes.length > 0) {
+    if (users.has(req.body.name) && users.get(req.body.name).points >= 10) {
+      const savedCode = users.get(req.body.name).code;
+      if (savedCode !== undefined) {
+        res.statusCode = 200;
+        res.send(savedCode);
+        return;
+      }
+      const code = giftcodes.pop();
+      saveCode(req.body.name, code);
+      res.statusCode = 200;
+      res.send(code);
+    } else {
+      res.statusCode = 403;
+      res.send("not enough points");
+    }
+  } else {
+    res.statusCode = 404;
```

Comment:
Maybe we can have a more specific error status than 404 for this, but we will discuss how the UI should handle if there are no more gift codes left.